### ISSUE TITLE
client-go / code-gen: enable client-gen/informer-gen/lister-gen to return user-provided Golang interfaces instead of versioned types

### ISSUE THREAD

**Issue by sttts** [kubernetes/kubernetes#56986]
Users of client-go often have to support multiple GroupVersions of an API in one operator. We do not support this use-case well in the generated clients.

As we do not offer internal versions, client-go users build abstractions over version types via Golang interfaces with getters and setters. Our generated clients are not prepared for this pattern, but potentially could. I am imagining some kind of user-provided shim on-top of versioned clients/informers/listers returning these user-provided interfaces. The code-generators could support this kind of pattern, i.e. by generating the boilerplate code around one-way conversions that the user provides.

**Comment by ncdc** [comment]
I haven't looked into this before. I don't really have a good answer atm. This may be tangentially related, but I do wonder if allowing a GroupVersion to add new resources or new fields to Kinds is also causing problems to some extent? If I'm a client and I ask for `apps/v1`, how do I know if that supports `Deployment.Spec.Frobble` (hypothetically added in Kubernetes v1.10)? Feel free to suggest moving this to another issue if you don't feel it's relevant to this one.

**Comment by sttts** [comment]
I think our only answer to new fields is to ask the server for its version today (which might be good enough for controllers). The json/protobuf (de)serializers will just ignore that field.

**Comment by ncdc** [comment]
True, but if you're an external user, how do you know that the absence of some "new" field means it wasn't set vs the server doesn't support it yet? There are different meanings of `apps/v1` (just an example) depending on kube version...

**Comment by sttts** [comment]
Right. There were discussions about subversions. But I think they will not solve anything we don't have now already with the server version. So you have to put some meta knowledge about `app/v1` into your controller logic. Luckily seeing a 1.9 server mostly means the same for every Kube flavor (maybe with the exception OpenShift which backports features). 

Feature flags might be another question. We cannot query them and fields might be rejected by validation. How is this handled today?

**Comment by caesarxuchao** [comment]
cc @cheftako

**Comment by jennybuckley** [comment]
cc @jennybuckley

**Comment by brancz** [comment]
I agree that the problem @ncdc mentioned would exist, however, it already exists today in the same way, as if you have an outdated version of client-go, this means that objects are unmarshaled with the available fields. This problem has occurred multiple times for us while developing the Prometheus Operator, however, seem orthogonal to the problem of not having to rewrite versions all the time in order to update client-go. Enabling to return the interfaces would ease at least one of these problems.

**Comment by ncdc** [comment]
Yeah, like I said, not trying to muddy the waters here unnecessarily :-)

@sttts can you shed some more light (pseudocode?) on how you'd envision this working?

**Comment by sttts** [comment]
@ncdc will try to do a very basic prototype, better than hand waving.
